solr/core/src/java/org/apache/solr/handler/admin/api/CoreAdminAPIBase.java

```diff
@@ -0,0 +1,157 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.admin.api;
+
+import static org.apache.solr.handler.admin.CoreAdminHandler.CoreAdminAsyncTracker.COMPLETED;
+import static org.apache.solr.handler.admin.CoreAdminHandler.CoreAdminAsyncTracker.FAILED;
+import static org.apache.solr.handler.admin.CoreAdminHandler.CoreAdminAsyncTracker.RUNNING;
+
+import java.util.function.Supplier;
+import org.apache.solr.api.JerseyResource;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.handler.admin.CoreAdminHandler;
+import org.apache.solr.handler.api.V2ApiUtils;
+import org.apache.solr.jersey.SolrJerseyResponse;
+import org.apache.solr.logging.MDCLoggingContext;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.util.tracing.TraceUtils;
+import org.slf4j.MDC;
+
+/**
+ * A common parent for admin Core Jersey-based APIs.
+ *
+ * <p>This base class is used when creating Core APIs to allow extra bookkeeping tasks such as async
+ * requests handling.
+ */
+public abstract class CoreAdminAPIBase extends JerseyResource {
+
+  protected final CoreContainer coreContainer;
+  protected final CoreAdminHandler.CoreAdminAsyncTracker coreAdminAsyncTracker;
+  protected final SolrQueryRequest req;
+  protected final SolrQueryResponse rsp;
+
+  public CoreAdminAPIBase(
+      CoreContainer coreContainer,
+      CoreAdminHandler.CoreAdminAsyncTracker coreAdminAsyncTracker,
+      SolrQueryRequest req,
+      SolrQueryResponse rsp) {
+    this.coreContainer = coreContainer;
+    this.coreAdminAsyncTracker = coreAdminAsyncTracker;
+    this.req = req;
+    this.rsp = rsp;
+  }
+
+  /**
+   * Wraps the subclasses logic with extra bookkeeping logic.
+   *
+   * <p>This method currently exists to enable async handling behavior for V2 Core APIs.
+   *
+   * <p>Since the logic for a given API lives inside the Supplier functional interface, checked
+   * exceptions can't be thrown directly to the calling method. To throw a checked exception out of
+   * the Supplier, wrap the exception using {@link CoreAdminAPIBase.CoreAdminAPIBaseException} and
+   * throw it instead. This handle method will retrieve the checked exception from {@link
+   * CoreAdminAPIBaseException} and throw it to the original calling method.
+   *
+   * @param solrJerseyResponse the response that the calling methods expects to return.
+   * @param coreName the name of the core that work is being done against.
+   * @param taskId an id provided for registering async work.
+   * @param actionName a name for the action being done.
+   * @param supplier the work that the calling method wants done.
+   * @return the supplied T solrJerseyResponse
+   */
+  public <T extends SolrJerseyResponse> T handle(
```

Comment:
[+1] Overall I really like how this approach panned out; very cool!

[0] Small nit: can you think of a name here that'd make the method's purpose clearer than `handle`?  I'm struggling a bit, but 'handle' feels generic. Maybe `handlePotentiallyAsynchronousTask`?

[-0] The one downside of this approach is that CoreAdminHandler still has a lot of code that looks similar to what's in this method.  (Specifically lines 200-250ish in CoreAdminHandler.)  Maybe this is as good as it gets, but can you think of any way we could refactor CoreAdminHandler, or this method, or both to cut down on the duplication between the two?